client.go

```diff
@@ -0,0 +1,318 @@
+package iprepd
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+)
+
+// Client is the iprepd service client
+type Client struct {
+	hostURL    string
+	authStr    string
+	httpClient *http.Client
+}
+
+// in an effort to keep the client's error handling consistent and for ease of
+// testing, all error messages should be listed below and reused where possible
+const (
+	// input validation errors
+	clientErrURLEmpty            = "url cannot be empty"
+	clientErrAuthEmpty           = "auth credentials cannot be empty"
+	clientErrObjectEmpty         = "object cannot be empty"
+	clientErrObjectTypeEmpty     = "object type cannot be empty"
+	clientErrViolationEmpty      = "violation cannot be empty"
+	clientErrReputationNil       = "reputation cannot be nil"
+	clientErrViolationRequestNil = "violation request cannot be nil"
+	clientErrMarshal             = "could not marshal payload"
+	// http client errors
+	clientErrBuildRequest = "could not build http request"
+	clientErrSendRequest  = "could not send http request"
+	// http response payload errors
+	clientErrReadResponse = "could not read response body"
+	clientErrNon200       = "non 200 status code received"
+	clientErrUnmarshal    = "could not unmarshal response body"
+)
+
+// NewClient is the default constructor for the client
+func NewClient(url, token string, httpClient *http.Client) (*Client, error) {
+	if url == "" {
+		return nil, errors.New(clientErrURLEmpty)
+	}
+	if token == "" {
+		return nil, errors.New(clientErrAuthEmpty)
+	}
+	if httpClient == nil {
+		httpClient = http.DefaultClient
+	}
+	return &Client{
+		hostURL:    url,
+		authStr:    token,
+		httpClient: httpClient,
+	}, nil
+}
+
+func (c *Client) addAuth(r *http.Request) {
+	r.Header.Set("Authorization", c.authStr)
+}
+
+// Dump retrieves all reputation entries
+func (c *Client) Dump() ([]Reputation, error) {
+	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("%s/dump", c.hostURL), nil)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %s", clientErrBuildRequest, err)
+	}
+	c.addAuth(req)
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %s", clientErrSendRequest, err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("%s: %d", clientErrNon200, resp.StatusCode)
+	}
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	defer resp.Body.Close()
+	if err != nil {
+		return nil, fmt.Errorf("%s: %s", clientErrReadResponse, err)
+	}
+	var ret []Reputation
+	if err = json.Unmarshal(bodyBytes, &ret); err != nil {
+		return nil, fmt.Errorf("%s: %s", clientErrUnmarshal, err)
+	}
+	return ret, nil
+}
+
+// Heartbeat checks whether an IPrepd deployment is healthy / reachable
+func (c *Client) Heartbeat() (bool, error) {
```

Comment:
Not sure if there is a ton of benefit in `Heartbeat` and `LBHeartbeat` being in here, the are used by the load balancers, not a big deal to leave them in though.

Reply:
yeah, i added them for completion of the API